src/VisualStudio/Core/Def/Implementation/ProjectSystem/Readme.md

```diff
@@ -0,0 +1,41 @@
+﻿# Locking and Synchronization
+
+There are two core types in this folder, each of which has their own lock.
+
+Each project from the project system is represented by a VisualStudioProject which has as SemaphoreSlim _gate. This lock is taken
+any time a mutation happens to that individual project, to ensure that the type itself is safe to use concurrently. The expectation
+though is that simultaneous use from multiple threads by a projet system isn't common, so no real effort has been expended to try
+to make the locking fine-grained there; each function just acquires the _gate and then does what it needs to do.
+
+The workspace (VisualStudioWorkspaceImpl) also has it's own SemaphoreSlim _gate. This is acquired any time a change is being made
+to the workspace. As much as possible, try to acquire this gate in an asynchronous fashion, since during solution load this lock
+will have a lot of things trying to acquire it at once on a bunch of threads, and we may starve off the thread pool if we're not
+careful. Unfortunately however, we still have legacy project systems that aren't async friendly; they still may apply changes or
+batches synchronously so in those cases we still acquire the gate in a synchronous fashion.
+
+There is a strict lock hierarchy: a VisualStudioProject may try to acquire the workspace lock while holding it's lock,
+but to prevent deadlocks a holder of the VisualStudioWorkspaceImpl lock should never call a function on VisualStudioProject that
+would acquire a project lock. To this end, a few bits of information that may seem to be "project specific" are actually stored
+in maps in the VisualStudioWorkspaceImpl; specifically we maintain a list of the output paths of a project which we use to convert
+metadata references to project references. This list is maintained in the workspace itself to avoid having to reach back to a
+project and ask it for information which might violate this lock hierarchy.
```

Comment:
this si very good to know.  i wonder if it's something we could trap somehow to detect when we're doing this wrong.